docs/auth/keycloak/keycloak-upstream-IAMs-conf.md

```diff
@@ -0,0 +1,107 @@
+# Integrate Upstream Identity Providers to Keycloak
+
+Keycloak has the ability to integrate with upstream identity providers. This way admins can use existing IAMs and /or users database pools while securing Tornjak. 
+
+![Upstream Architecture Diagram](diagrams/upstream-architecture-diagram.png)
+
+> [!NOTE]
+> As long as your upstream IAM uses `OAuth v2.0`, `OpenID Connect v1.0` or `SAML v2.0` you can integrate your IAM to keycloak. 
+
+This documentation is a guide on how to connect some hand picked IAMs. For more IAM connection options, go to your keycloak console and head to the `Identity providers` section. There you can see multple options including: 
+
+- Another Keycloak OpenID service
+- SAML
+- Google
+- Microsoft Azure - Active Directories (AD)
+- Openshift v3 and v4
+- Github
+- etc...
+
+![Identity Providers Homepage](diagrams/identity-providers-homepage.png)
+
+### Keycloak setup for Upstream IAMs
+The keycloak setup to connect any upstream IAM is pretty much standard. Please follow the steps below to connect your upsteeam IAM of choice. Go to keycloak console and select the IAM of interest as shown in the picture above. 
+
+> [!IMPORTANT] 
+> Make sure you are in the `tornjak` realm or the specific realm Tornjak app is registered on 
+- The `Redirect URI` is automatically set by keycloak, this is the uri you should enter in the `Redirect URI` section while creating/ updating your upstream IAM account (check below for specific upstream IAM instuctions). 
+- Fill in the `Client ID` and `Client Secret` generated in the upstream IAM in the respective fields. [Go to your upstream IAM account to get the `Client ID` and `Client Secret`]
+- For `Microsoft Azure`single-tenant auth endpoints, fill in the `Tenant ID`. If not specified uses 'common' multi-tenant endpoints. 
+- And click `Add`
+
+> [!IMPORTANT] Make sure you assign appropraite roles within keycloak for your user, or the roles are mapped correctly between the upstream IAM and keycloak. Check the Mappers section for more roles configuration. [****DETAILED MAPPERS SECTION TODO!!!****]
+
+> Now when you try signing in to the Tornjak application, you should see the keycloak login page and the upstream IAM as an optional upstream identity provider to sign in to. If you select the upstream IAM, keycloak will redirect you to sign in with the upstream IAM of choice and tornjak will be authenticated using that IAM. 
+
+### Microsoft
+![Microsoft Keycloak](diagrams/microsoft-keycloak.png)
```

Comment:
I do not have the microsoft keycloak Tenant ID here...

Reply:
Which Keycloak are you running? Is it through our examples or through helm charts install?